internal/merger/merger.go

```diff
@@ -26,65 +30,98 @@ import (
 )
 
 // MergeResolve merges two V2 resolve responses.
-func MergeResolve(r1, r2 *pbv2.ResolveResponse) *pbv2.ResolveResponse {
-	if r1 == nil {
-		return r2
-	} else if r2 == nil {
-		return r1
-	}
-
-	// Maps are used to dedup: node -> resolved ID -> candidate.
-	store := map[string]map[string]*pbv2.ResolveResponse_Entity_Candidate{}
-
-	collectEntities := func(r *pbv2.ResolveResponse) {
-		for _, e := range r.GetEntities() {
-			node := e.GetNode()
-			if _, ok := store[node]; !ok {
-				store[node] = map[string]*pbv2.ResolveResponse_Entity_Candidate{}
-			}
-			for _, candidate := range e.GetCandidates() {
-				store[node][candidate.GetDcid()] = candidate
+func MergeResolve(main, aux *pbv2.ResolveResponse) *pbv2.ResolveResponse {
+	if main == nil {
+		return aux
+	}
+	if aux == nil {
+		return main
+	}
+	// Change aux list into map for easy lookup
+	auxStore := map[string]*pbv2.ResolveResponse_Entity{}
+	for _, e := range aux.GetEntities() {
+		node := e.Node
+		auxStore[node] = e
+	}
+	// Merge common entities from aux into main
+	for _, e := range main.GetEntities() {
+		node := e.Node
+		if auxEntity, ok := auxStore[node]; ok {
+			existCandidates := map[string]struct{}{}
+			for _, c := range e.Candidates {
+				existCandidates[c.Dcid] = struct{}{}
+			}
+			for _, c := range auxEntity.Candidates {
+				if _, ok := existCandidates[c.Dcid]; !ok {
+					e.Candidates = append(e.Candidates, c)
+				}
 			}
+			delete(auxStore, node)
 		}
 	}
-
-	collectEntities(r1)
-	collectEntities(r2)
-
-	res := &pbv2.ResolveResponse{}
-	for node, idToCandidate := range store {
-		candidates := []*pbv2.ResolveResponse_Entity_Candidate{}
-		for _, candidate := range idToCandidate {
-			candidates = append(candidates, candidate)
+	// Add aux entities that are not in main
+	for _, e := range aux.Entities {
+		if _, ok := auxStore[e.Node]; ok {
+			main.Entities = append(main.Entities, e)
 		}
-
-		// Sort to make result deterministic.
-		sort.Slice(candidates, func(i, j int) bool {
-			return candidates[i].GetDcid() < candidates[j].GetDcid()
-		})
-
-		res.Entities = append(res.Entities, &pbv2.ResolveResponse_Entity{
-			Node:       node,
-			Candidates: candidates,
-		})
 	}
+	return main
+}
 
-	// Sort to make result deterministic.
-	sort.Slice(res.Entities, func(i, j int) bool {
-		return res.Entities[i].Node < res.Entities[j].Node
-	})
-
-	return res
+func mergeLinkedGraph(
+	mainData, auxData map[string]*pbv2.LinkedGraph,
+) map[string]*pbv2.LinkedGraph {
```

Comment:
The triples API seems involved enough that we should probably have some focussed e2e tests for local vs. remote, in different cases (node-level, prop-level, PV-level, ...) ?  Do we have those? If not, have a TODO?

Reply:
There is a unit test for this function that should be good to guard but that does need more testing cases with real data. I will leave a todo and update the unittest.